src/ripple/app/tx/impl/SetOracle.cpp

```diff
@@ -0,0 +1,287 @@
+    // of the last closed ledger
+    using namespace std::chrono;
+    std::size_t const closeTime =
+        duration_cast<seconds>(ctx.view.info().closeTime.time_since_epoch())
+            .count();
+    std::size_t const lastUpdateTime = ctx.tx[sfLastUpdateTime];
+    if (lastUpdateTime <= epoch_offset.count())
+        return tecINVALID_UPDATE_TIME;
+    std::size_t const lastUpdateTimeEpoch =
+        lastUpdateTime - epoch_offset.count();
+    if (lastUpdateTimeEpoch < closeTime ||
+        lastUpdateTimeEpoch > (closeTime + maxLastUpdateTimeDelta))
+        return tecINVALID_UPDATE_TIME;
+
+    auto const sle = ctx.view.read(keylet::oracle(
+        ctx.tx.getAccountID(sfAccount), ctx.tx[sfOracleDocumentID]));
+
+    hash_set<uint256> pairs;
+    hash_set<uint256> pairsDel;
+    for (auto const& entry : ctx.tx.getFieldArray(sfPriceDataSeries))
+    {
+        auto const hash = tokenPairHash(entry);
+        if (pairs.contains(hash))
+            return tecDUPLICATE;
+        if (entry.isFieldPresent(sfAssetPrice))
+            pairs.emplace(hash);
+        else if (sle)
+            pairsDel.emplace(hash);
+        else
+            return temMALFORMED;
+    }
+
+    // Lambda is used to check if the value of a field, passed
+    // in the transaction, is equal to the value of that field
+    // in the on-ledger object.
+    auto isConsistent = [&ctx, &sle](auto const& field) {
+        auto const v = ctx.tx[~field];
+        return !v || *v == (*sle)[field];
+    };
+
+    if (sle)
+    {
+        // update
+        // Account is the Owner since we can get sle
+
+        // lastUpdateTime must be more recent than the previous one
+        if (ctx.tx[sfLastUpdateTime] <= (*sle)[sfLastUpdateTime])
+            return tecINVALID_UPDATE_TIME;
+
+        if (!isConsistent(sfProvider) || !isConsistent(sfAssetClass))
+            return temMALFORMED;
+
+        for (auto const& entry : sle->getFieldArray(sfPriceDataSeries))
+        {
+            auto const hash = tokenPairHash(entry);
+            if (!pairs.contains(hash))
+            {
+                if (pairsDel.contains(hash))
+                    pairsDel.erase(hash);
+                else
+                    pairs.emplace(hash);
+            }
+        }
+        if (!pairsDel.empty())
+            return tecOBJECT_NOT_FOUND;
+    }
+    else
+    {
+        // create
+
+        if (!ctx.tx.isFieldPresent(sfProvider) ||
+            !ctx.tx.isFieldPresent(sfAssetClass))
+            return temMALFORMED;
+    }
+
+    if (pairs.empty())
+        return temARRAY_EMPTY;
+    if (pairs.size() > maxOracleDataSeries)
+        return temARRAY_TOO_LARGE;
+
+    auto const add = pairs.size() > 5 ? 2 : 1;
+    auto const reserve = ctx.view.fees().accountReserve(
+        sleSetter->getFieldU32(sfOwnerCount) + add);
+    auto const& balance = sleSetter->getFieldAmount(sfBalance);
+
+    if (balance < reserve)
+        return tecINSUFFICIENT_RESERVE;
```

Comment:
Ditto with the reserve check. Looks like in other transactions it's always just done in `doApply`.

Reply:
Well, strictly speaking there is a precedence for the reserve check in `preclaim`. For instance, `PayChan`, `XChainBridge`, and `Payment`. Again, I want to do as much as possible of the transaction parameters related error checking outside of apply. I can move to apply if you feel strongly about it.